mailpoet/lib/Migrations/App/Migration_20240207_105912_App.php

```diff
@@ -0,0 +1,173 @@
+
+    // update sending queue counts
+    $this->entityManager->createQueryBuilder()
+      ->update(SendingQueueEntity::class, 'q')
+      ->set('q.countProcessed', 'q.countTotal')
+      ->set('q.countToProcess', 0)
+      ->where('q.task IN (:ids)')
+      ->setParameter('ids', $ids)
+      ->getQuery()
+      ->execute();
+
+    // complete the invalid tasks
+    $this->entityManager->createQueryBuilder()
+      ->update(ScheduledTaskEntity::class, 't')
+      ->set('t.status', ':completed')
+      ->where('t.id IN (:ids)')
+      ->setParameter('completed', ScheduledTaskEntity::STATUS_COMPLETED)
+      ->setParameter('ids', $ids)
+      ->getQuery()
+      ->execute();
+
+    // mark newsletters as sent, update "sentAt" (DBAL needed to be able to use JOIN)
+    $newslettersTable = $this->entityManager->getClassMetadata(NewsletterEntity::class)->getTableName();
+    $scheduledTasksTable = $this->entityManager->getClassMetadata(ScheduledTaskEntity::class)->getTableName();
+    $sendingQueuesTable = $this->entityManager->getClassMetadata(SendingQueueEntity::class)->getTableName();
+    $this->entityManager->getConnection()->executeStatement(
+      "
+        UPDATE $newslettersTable n
+        JOIN $sendingQueuesTable q ON n.id = q.newsletter_id
+        JOIN $scheduledTasksTable t ON q.task_id = t.id
+        SET
+          n.status = :sent,
+          n.sent_at = t.updated_at
+        WHERE t.id IN (:ids)
+      ",
+      ['sent' => NewsletterEntity::STATUS_SENT, 'ids' => $ids],
+      ['ids' => Connection::PARAM_INT_ARRAY]
+    );
+  }
+
+  private function backfillMissingDataForMigratedNewsletters(): void {
+    // In https://mailpoet.atlassian.net/browse/MAILPOET-5886 we fixed missing "sent" status
+    // by https://github.com/mailpoet/mailpoet/pull/5416, but didn't backfill missing data.
+
+    // get affected newsletter IDs
+    $ids = $this->entityManager->createQueryBuilder()
+      ->select('n.id')
+      ->from(NewsletterEntity::class, 'n')
+      ->where('n.status = :sent')
+      ->andWhere('n.sentAt IS NULL')
+      ->setParameter('sent', NewsletterEntity::STATUS_SENT)
+      ->getQuery()
+      ->getSingleColumnResult();
+
+    // get missing newsletter statistics IDs
+    $data = $this->entityManager->createQueryBuilder()
+      ->select('IDENTITY(q.newsletter) AS nid, q.id AS qid, IDENTITY(s.subscriber) AS sid, s.updatedAt AS sentAt')
+      ->from(SendingQueueEntity::class, 'q')
+      ->join('q.task', 't')
+      ->join('t.subscribers', 's')
+      ->leftJoin(StatisticsNewsletterEntity::class, 'ns', 'WITH', 'ns.queue = q AND ns.subscriber = s.subscriber')
+      ->where('q.newsletter IN (:ids)')
+      ->andWhere('ns.id IS NULL')
+      ->andWhere('s.processed = :processed')
+      ->andWhere('s.failed = :ok')
+      ->setParameter('ids', $ids)
+      ->setParameter('processed', ScheduledTaskSubscriberEntity::STATUS_PROCESSED)
+      ->setParameter('ok', ScheduledTaskSubscriberEntity::FAIL_STATUS_OK)
```

Comment:
I suppose we save newsletter statistics only for those, as the entity has a `sentAt` field.

Reply:
Actually, we don't, if I understand it correctly here: https://github.com/mailpoet/mailpoet/blob/bbe471e6531a1e821020097303d89f557c7325b2/mailpoet/lib/Cron/Workers/SendingQueue/SendingQueue.php#L512-L543

In L521 we check the `sendResult['response']` and process errors, but as we do not return, we will create the entries in L543.

Reply:
@websupporter Ahh, right. So they must be `processed` (otherwise they wouldn't get to that stats saving logic), but `failed` status we seem to ignore.